### DISCUSSION TITLE
How to set multiplatform flavor based on XCode target with CocoaPods?

### DISCUSSION THREAD

**Discussion by mr-kew** [yshrsmz/BuildKonfig#151] [Q&A] [answered]
I used to use script build phase for this on kotlin version `1.19.20` and BuildKonfig version `1.5.0` and it worked fine:
```
if [ $ENABLE_PREVIEWS = "YES" ]; then exit 0; fi
cd "$SRCROOT/.." && ./gradlew :shared:embedAndSignAppleFrameworkForXcode -P buildkonfig.flavor=prod
```
It is somewhat inspired by [the futured template](https://github.com/futuredapp/kmp-futured-template)

But since updating to kotlin to version `2.0.0` and `BuildKonfig` version 1.5.1, it does not compile and says:
```
'embedAndSign' task can not be used in a project with dependencies to pods.
```
I tried switching to any other task like `build`, `assemble` or `generateBuildKonfig`, and moving the build phase up and down the list. It compiles & runs, but the flavor is not set.

Upon reverting back to original versions it turns out that the flavor is also not set when using other tasks instead of `embedAndSignAppleFrameworkForXcode`. 

### So the question is:

What other task can I use to set the multiplatform flavor from XCode target's build phases instead since the `embedAndSignAppleFrameworkForXcode` is now prohibited for CocoaPods?

**Comment by mr-kew** [comment, marked as answer]
Ok, so I finally figured it out.

**Running `embedAndSign` task in a CocoaPods project to set a `gradle.property` is wrong.** As discussed in [the original issue][1], it can lead to linking issues and generally does not make much sense, as the `shared` module is already build as a pod.

Intended way (as I understand it):

 1. Edit iOS configurations to contain the flavor name (e.g. `ProdDebug`). It can be done in project settings under the Info tab. CocoaPods should then automatically create corresponding configuration files under the `Pods` folder.

 2. Edit your scheme for each target to run the corresponding configuration.

 3. The configuration name is then automatically passed to the gradle when building the module inside the `shared.podspec`:
```
"$REPO_ROOT/../gradlew" -p "$REPO_ROOT" $KOTLIN_PROJECT_PATH:syncFramework \
    -Pkotlin.native.cocoapods.platform=$PLATFORM_NAME \
    -Pkotlin.native.cocoapods.archs="$ARCHS" \
    -Pkotlin.native.cocoapods.configuration="$CONFIGURATION"
```

 4. Map new iOS configurations to corresponding `NativeBuildType`s:
 ```
xcodeConfigurationToNativeBuildType["FlavorDebug"] = NativeBuildType.DEBUG
xcodeConfigurationToNativeBuildType["FlavorRelease"] = NativeBuildType.RELEASE
```

 5. Access the configuration name:
```
val configuration = findProperty(KotlinCocoapodsPlugin.CONFIGURATION_PROPERTY)
```

 6. Extract the flavor from it and set the multiplatform flavor before the `generateBuildKonfig` task, using `afterEvaluate ` block for example:
```
afterEvaluate {
    setProperty("buildkonfig.flavor", flavor)
}
```

  [1]: https://youtrack.jetbrains.com/issue/KT-64096/Diagnostic-when-embedAndSign-used-for-framework-with-cocoapods-dependencies

**Reply by Qw4z1** [reply]
Wait... You lost me at 5. =) You wouldn't happen have a small sample project with this set up, would you?

**Reply by mr-kew** [reply]
@Qw4z1 Sadly no. But you should do 5 inside the `afterEvaluate` block. It should return String containing name of your configuration.
